Text/HealthcareTA/HealthcareTA.cs

```diff
@@ -0,0 +1,136 @@
+            ILogger log,
+            ExecutionContext executionContext)
+        {
+            string skillName = executionContext.FunctionName;
+            IEnumerable<WebApiRequestRecord> requestRecords = WebApiSkillHelpers.GetRequestRecords(req);
+            if (requestRecords == null)
+            {
+                return new BadRequestObjectResult($"{skillName} - Invalid request record array.");
+            }
+
+            // Get Endpoint and access key from App Settings
+            string apiKey = Environment.GetEnvironmentVariable(healthcareApiEnvKey, EnvironmentVariableTarget.Process);
+            string apiEndpoint = Environment.GetEnvironmentVariable(healthcareApiEnvEndpoint, EnvironmentVariableTarget.Process);
+            if (apiKey == null || apiEndpoint == null)
+            {
+                return new BadRequestObjectResult($"{skillName} - Healthcare Text Analytics API key or endpoint is missing. Make sure to set them in the Environment Variables.");
+            }
+
+            var client = new TextAnalyticsClient(new Uri(apiEndpoint), new AzureKeyCredential(apiKey));
+
+            // Get a custom timeout from the header, if it exists. If not use the default timeout.
+            int timeout;
+            if (!int.TryParse(req.Headers["SkillTimeout"].ToString(), out timeout))
+            {
+                timeout = defaultTimeout;
+            }
+            var timeoutMiliseconds = timeout * 1000;
+
+            WebApiSkillResponse response = await WebApiSkillHelpers.ProcessRequestRecordsAsync(skillName, requestRecords,
+                async (inRecord, outRecord) => {
+                    // Prepare analysis operation input
+                    var document = inRecord.Data["document"] as string;
+                    var options = new AnalyzeHealthcareEntitiesOptions { };
+                    if (document.Length >= maxCharLength)
+                    {
+                        outRecord.Warnings.Add(new WebApiErrorWarningContract
+                        {
+                            Message = $"Healthcare Text Analytics Error: The submitted document was over {maxCharLength} characters. It has been truncated to fit this requirement."
+                        });
+                        document = document.Substring(0, maxCharLength);
+                    }
+                    List<string> batchInput = new List<string>()
+                    {
+                        document
+                    };
+
+                    // start analysis process
+                    var timer = System.Diagnostics.Stopwatch.StartNew();
+                    AnalyzeHealthcareEntitiesOperation healthOperation = await client.StartAnalyzeHealthcareEntitiesAsync(batchInput, "en", options);
+                    var healthOperationTask = healthOperation.WaitForCompletionAsync().AsTask();
+
+                    if (await Task.WhenAny(healthOperationTask, Task.Delay(timeoutMiliseconds)) == healthOperationTask)
```

Comment:
>if (await Task.WhenAny(healthOperationTask, Task.Delay(timeoutMiliseconds)) == healthOperationTask) [](http://example.com/codeflow?start=20&length=99)

So keep in mind a feww things here:

1. That max timeout of 230 seconds is the max overall function runtime. So if you wait until now to start that timer, it'll probably be too late and the function will just crash without sending the nice error message we want. We should probably add a 5 seconds buffer or something to counteract that.
2. You are measuring the timeout per request, but you might be sending multiple requests per input depending on how the customer configured the skill, and as I mentioned the timeout needs to be for the function in its entirety.

Reply:
Updated to check for timeout across all requests